### PR TITLE
JWT Assertion login for CLI 

### PR DESCRIPTION
**Pull request by jamestouri**
https://github.com/medplum/medplum/issues/2657

Ran:
```
ts-node ~/Developer/medplum/packages/cli/src/index.ts login -p the-assertion \
—-authorize-url "https://fhir.epic.com/interconnect-fhir-oauth/oauth2/authorize" \
—-token-url “oauth2/token" \
—-client-id "{CLIENT_ID}" \
--fhir-url-path "fhir/R4"\
--private-key-path "{/PRIVATEKEY/PATH.PEM}"\
--base-url "https://fhir.epic.com/interconnect-fhir-oauth/" \
--auth-type jwt-assertion \
--audience "oauth2/token"
```
Screenshot 2023-08-09 at 8 40 38 AM

### CODE REVIEW

File: packages/docs/docs/cli/external-fhir-servers.md [hunk 1]
```diff
@@ -85,6 +85,25 @@ medplum profile set example \
...
+    --fhir-url-path "fhir/R4" \
+    --token-url "/oauth2/token" \
+    --client-id "MY_CLIENT_ID" \
+    --audience "/oauth2/token" \
+    --subject "john_doe" \
+    --private-key-path "/path/to/privatekey.pem"
+```
+
+:::note
+If you are authenticating with Epic and it returns {"error":"invalid_client","error_description":null}, here is a [Doc to Troubleshoot](https://fhir.epic.com/Resources/jwt_auth_troubleshoot_eof).
```
Comment (jamestouri) [inline packages/docs/docs/cli/external-fhir-servers.md:103]: Added the troubleshoot doc here

File: packages/cli/src/auth.ts [hunk 2]
```diff
@@ -35,21 +34,27 @@ whoami.action(async (options) => {
   printMe(medplum);
 });
 
-async function startLogin(medplum: MedplumClient, profile: Profile): Promise<void> {
+async function startLogin(medplum: MedplumClient, profile?: Profile): Promise<void> {
```
Comment (codyebberson) [inline packages/cli/src/auth.ts:37]: What is the case where `profile` is undefined?

File: packages/cli/src/utils.ts [hunk 3]
```diff
@@ -1,10 +1,13 @@
...
 import { Bot, Extension, OperationOutcome } from '@medplum/fhirtypes';
 import { existsSync, readFileSync, writeFile } from 'fs';
-import { basename, extname, resolve } from 'path';
+import { basename, extname, resolve, join } from 'path';
 import internal from 'stream';
 import tar from 'tar';
 import { FileSystemStorage } from './storage';
+import { SignJWT } from 'jose';
+import { createPrivateKey, randomBytes, createHmac } from 'crypto';
+import { homedir } from 'os';
```
Comment (codyebberson) [inline packages/cli/src/utils.ts:10]: Please run "Organize imports" in VS Code

image

(Note... we should make this a linting rule somehow)

**Review by codyebberson — COMMENTED** [review]
Nice!  This will be a great addition!

File: packages/cli/src/auth.ts (hunk 2 above)
Comment (jamestouri) [reply packages/cli/src/auth.ts:37]: running `medplum login` will have profile as undefined

Comment (jamestouri) [reply packages/cli/src/auth.ts:37]: Good point, it's never

**Review by codyebberson — APPROVED** [review]

**Pull request merged**